test/gurka/test_recost.cc

```diff
@@ -0,0 +1,309 @@
+        auto leg = api.trip().routes(0).legs(0);
+
+        // setup a callback for the recosting to get each edge
+        auto edge_itr = leg.node().begin();
+        sif::EdgeCallback edge_cb = [&edge_itr]() -> baldr::GraphId {
+          auto edge_id =
+              edge_itr->has_edge() ? baldr::GraphId(edge_itr->edge().id()) : baldr::GraphId{};
+          ++edge_itr;
+          return edge_id;
+        };
+
+        // setup a callback for the recosting to tell us about the new label each made
+        auto elapsed_itr = leg.node().begin();
+        double length = 0;
+        uint32_t pred = baldr::kInvalidLabel;
+        sif::LabelCallback label_cb = [&elapsed_itr, &length,
+                                       &pred](const sif::EdgeLabel& label) -> void {
+          length += elapsed_itr->edge().length() * 1000.0;
+          EXPECT_EQ(elapsed_itr->edge().id(), label.edgeid());
+          EXPECT_EQ(pred++, label.predecessor());
+          EXPECT_EQ(static_cast<uint8_t>(elapsed_itr->edge().travel_mode()),
+                    static_cast<uint8_t>(label.mode()));
+          EXPECT_NEAR(length, label.path_distance(), 2);
+          EXPECT_NEAR(elapsed_itr->cost().transition_cost().seconds(), label.transition_cost().secs,
+                      .1);
+          // TODO: test restrictions
+          // we need to move to the next node which has the elapsed time at the end of the edge
+          ++elapsed_itr;
+          EXPECT_NEAR(elapsed_itr->cost().elapsed_cost().seconds(), label.cost().secs, .1);
+        };
+
+        // find the percentage of the edges used
+        float src_pct = 0;
+        for (const auto& edge : api.options().locations(0).path_edges()) {
+          if (leg.node(0).edge().id() == edge.graph_id()) {
+            src_pct = edge.percent_along();
+            break;
+          }
+        }
+        float tgt_pct = 1;
+        for (const auto& edge : api.options().locations(1).path_edges()) {
+          if (std::next(leg.node().rbegin())->edge().id() == edge.graph_id()) {
+            tgt_pct = edge.percent_along();
+            break;
+          }
+        }
+
+        // is there time dependence and in what direction
+        auto dt_itr = option.find("/date_time/value");
+        std::string date_time = dt_itr != option.cend() ? dt_itr->second : "";
+        auto type_itr = option.find("/date_time/type");
+        bool forward = type_itr != option.cend() ? std::stoi(type_itr->second) != 2 : true;
+        // build up the costing object
+        auto costing = sif::CostFactory().Create(api.options());
+
+        // recost the path
+        if (forward) {
+          sif::recost_forward(*reader, *costing, edge_cb, label_cb, src_pct, tgt_pct, date_time);
+        } else {
+          // sif::recost_reverse(*reader, &mode_costing[0], date_time, edge_cb, label_cb);
+        }
+      }
+    }
+  }
+}
+
+TEST(recosting, throwing) {
+  const std::string ascii_map = R"(A--1--B-2-3-C
```

Comment:
this one checks that recosting fails for various scenarios of starting on an edge where your recost isnt allowed or having edges in the middle that arent allowed or giving bad input to the recosting method and so on